File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    /* two pass solution
// create a dummy node for deleting a starting index
    let sentinel = new ListNode();
    sentinel.next = head;

    // calculate linkedlist length;

    let length=0;
    while(head){
        head=head.next;
        length++;
    }

    let prevPos = length - n;

    let prev = sentinel;
    for(let i=0; i<prevPos;i++){
        prev=prev.next;
    }
    prev.next=prev.next.next;
    return sentinel.next;*/

    let sentinel = new ListNode();
    sentinel.next= head;

    let fast = sentinel;
    let slow = sentinel;

    for(let i=0;i<=n;i++){
        fast=fast.next;
    }

    while(fast!=null){
        fast=fast.next;
        slow=slow.next;
    }
    slow.next = slow.next.next;

    return sentinel.next;

   
    
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
/*
        ListNode slow = head;
        ListNode fast = head;


        for(int i=0;i<n;i++){
            fast=fast.next; // slow=head, fast=3
        }

        if(fast==null){
            return head.next;
        }

        while(fast.next!=null){
            slow=slow.next;
            fast=fast.next;
        }
        slow.next=slow.next.next;
        return head;*/

        ListNode sentinel = new ListNode();
        sentinel.next = head;

        ListNode first = sentinel;
        for(int i=0;i<=n;i++){
            first=first.next;
        }
        ListNode second = sentinel;
        while(first!=null){
            first=first.next;
            second=second.next;
        }
        second.next=second.next.next;
        return sentinel.next;


        
        
        
    }
}